pkg/sql/subquery.go

```diff
@@ -218,16 +216,15 @@ func (v *subqueryPlanVisitor) enterNode(_ string, n planNode) bool {
 	return true
 }
 
-func (v *subqueryPlanVisitor) attr(_, _, _ string)                    {}
-func (v *subqueryPlanVisitor) expr(_, _ string, _ int, _ parser.Expr) {}
-func (v *subqueryPlanVisitor) leaveNode(_ string)                     {}
-
 func (p *planner) startSubqueryPlans(plan planNode) error {
 	// We also run and pre-evaluate the subqueries during start,
 	// so as to avoid re-running the sub-query for every row
 	// in the results of the surrounding planNode.
 	p.subqueryPlanVisitor = subqueryPlanVisitor{p: p}
-	return walkPlan(plan, &p.subqueryPlanVisitor)
+	return walkPlan(plan, planObserver{
```

Comment:
> and drop the checks on nil in visit() to increase readability.

The nil checks in `planVisitor.visit` are the primary reason for this PR. In particular, notice how we don't generate the `attr` field to pass to `planObserver.attr` unless that field is non-nil.